services/src/pro/datasets/storage.rs

```diff
@@ -0,0 +1,197 @@
+                deletion_timestamp: None,
+                delete_record: false,
+                delete_data: true,
+            }),
+        }
+    }
+
+    pub fn delete_full(dataset_id: DatasetId) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: None,
+                delete_record: true,
+                delete_data: true,
+            }),
+        }
+    }
+
+    pub fn delete_none(dataset_id: DatasetId) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: None,
+                delete_record: false,
+                delete_data: false,
+            }),
+        }
+    }
+
+    pub fn expire_meta(dataset_id: DatasetId, timestamp: DateTime) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: Some(timestamp),
+                delete_record: true,
+                delete_data: false,
+            }),
+        }
+    }
+
+    pub fn expire_fair(dataset_id: DatasetId, timestamp: DateTime) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: Some(timestamp),
+                delete_record: false,
+                delete_data: true,
+            }),
+        }
+    }
+
+    pub fn expire_full(dataset_id: DatasetId, timestamp: DateTime) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: Some(timestamp),
+                delete_record: true,
+                delete_data: true,
+            }),
+        }
+    }
+
+    pub fn expire_none(dataset_id: DatasetId, timestamp: DateTime) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::SetExpire(Expiration {
+                deletion_timestamp: Some(timestamp),
+                delete_record: false,
+                delete_data: false,
+            }),
+        }
+    }
+
+    pub fn unset_expire(dataset_id: DatasetId) -> Self {
+        ChangeDatasetExpiration {
+            dataset_id,
+            expiration_change: ExpirationChange::UnsetExpire,
+        }
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, ToSchema)]
+#[serde(rename_all = "camelCase", tag = "type")]
+pub enum ExpirationChange {
+    SetExpire(Expiration),
+    UnsetExpire,
+}
+
+pub enum UploadedDatasetStatus {
+    Available,
+    Deleted,
+}
+
+#[derive(Debug, FromSql, ToSql)]
+pub enum InternalUploadedDatasetStatus {
+    Available,
+    Expires,
+    Expired,
+    UpdateExpired,
```

Comment:
what does this state mean?

Reply:
Available: Dataset is available.
Expires: Dataset was set to expire at some point in the future or dataset was "deleted", which translates to expire now.
Expired: Dataset has reached the expiration timestamp, and the deletion action can be triggered now.
UpdateExpired: Dataset was previously Expired or Deleted, and the user has "upgraded" this expiration/deletion. The only way now, after incorporating the other review comments, is that the metadata should also be deleted. Previously this covered other cases. Will be marked as Expired once done.
Deleted: Action after Expired was taken.

Reply:
maybe add it as a comment to the code